feature airgap for downstream-cluster (#2)
suse-edge/atip: telco-examples/edge-clusters/airgap/README.md

* In the following examples, we are assuming that the edge cluster will use Baremetal Servers. If you want to deploy the full workflow using virtual machines, please refer to the [metal3-demo repo](https://github.com/suse-edge/metal3-demo)
* In the following examples, as we're creating an air-gap scenario, we are assuming that you have a local private registry deployed in the same network. 

## Create the image for the edge cluster

### Prerequisites

To prepare an airgap environment, we will use a private local registry (already deployed and configured) to store some images (helm-chart oci images and some other images), and we will use the Edge Image Builder to include the rke2-images artifacts which are a requirement to use the cluster-api provider for rke2.

Using the example folder `telco-examples/edge-clusters/airgap/eib` we will create the basic structure in order to build the image for the edge cluster: 

You need to modify the following values in the `telco-edge-airgap-cluster.yaml` file:

> - `${ROOT_PASSWORD}` - The root password for the management cluster. This could be generated using `openssl passwd -6 PASSWORD` and replacing PASSWORD with the desired password, and then replacing the value in the `telco-edge-cluster.yaml` file.

Author: ohh good catch!!! copy&paste from mgmt-cluster. Thanks

- `${SCC_REGISTRATION_CODE}` - The registration code for the SUSE Customer Center for the SLE Micro product. This could be obtained from the SUSE Customer Center and replacing the value in the `telco-edge-cluster.yaml` file.

You need to modify the following folder:

- `base-images` - To include inside the `SLE-Micro.x86_64-5.5.0-Default-RT-GM.raw` image downloaded from the SUSE Customer Center.

> ### preparing the airgap aritfacts

Author: also another typo in artifacts word. Thanks

The following steps are required to prepare the airgap artifacts:

1.Include the rke2-release-images to the `custom/files` folder to be consumed by EIB during the build process. 
  - You can use the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-save-rke2-images.sh) and the list of images [here](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-release-rke2-images.txt) to generate the artifacts required to be included in `custom/files`. 
  ```
  $ ./edge-save-rke2-images.sh -o ~/telco-examples/edge-clusters/airgap/eib/custom/files -l ~/edge-release-rke2-images.txt
  ...
  $ tree ~/telco-examples/edge-clusters/airgap/eib/custom/files
  .
  |-- install.sh
  |-- rke2-images-cilium.linux-amd64.tar.zst
  |-- rke2-images-core.linux-amd64.tar.zst
  |-- rke2-images-multus.linux-amd64.tar.zst
  |-- rke2-images.linux-amd64.tar.zst
  |-- rke2.linux-amd64.tar.gz
  `-- sha256sum-amd64.txt
  ```

> 2. Preload your registry with the helm-chart oci images required for the edge cluster. 

Reviewer: ```suggestion
2. Prepare your artifact registry.
```

Author: mmm I'd like to keep the preload. it's really common when you're working with disconnected environments

>   - You need to create a list with the helm-chart oci images required for the edge cluster. For example, for telco scenarios,  you can use the following list:

Reviewer: ```suggestion
  - You need to create a list with the Helm charts required for the edge cluster. For example, for telco scenarios, you can use the following list:
```

    ``` 
    $ cat > edge-release-helm-oci-artifacts.txt <<EOF
    edge/sriov-network-operator-chart:1.2.2
    edge/sriov-crd-chart:1.2.2
    EOF
    ```

>   - Using the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-save-oci-artefacts.sh) and the list created above, you can generate in local the tarball with the helm-chart oci images required for the edge cluster.

Reviewer: ```suggestion
  - Using the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-save-oci-artefacts.sh) and the list created above, you can generate a tarball containing all necessary Helm charts locally.
```

Author: much better. Thanks

    ```
    $ ./edge-save-oci-artefacts.sh -al ./edge-release-helm-oci-artifacts.txt -s registry.suse.com
    Pulled: registry.suse.com/edge/sriov-network-operator-chart:1.2.2
    Pulled: registry.suse.com/edge/sriov-crd-chart:1.2.2
    a edge-release-oci-tgz-20240705
    a edge-release-oci-tgz-20240705/sriov-network-operator-chart-1.2.2.tgz
    a edge-release-oci-tgz-20240705/sriov-crd-chart-1.2.2.tgz
    ```

>   - Upload your tarball to your private registry to preload with the helm chart oci images downloaded in the previous step using the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-load-oci-artefacts.sh):

Reviewer: ```suggestion
  - Upload your tarball to your private registry to load the downloaded Helm charts using the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-load-oci-artefacts.sh):
```

    ```
    $ tar zxvf edge-release-oci-tgz-20240705.tgz
    $ ./edge-load-oci-artefacts.sh -ad edge-release-oci-tgz-20240705 -r myregistry:5000
    ```

> 3. Preload your registry with the rest of the images as well as your workload images required for the edge cluster. 

Reviewer: ```suggestion
3. Preload your registry with the necessary container images (incl. your workload ones) required for the edge cluster. 
```

    a edge-release-images-tgz-20240705/hardened-sriov-network-config-daemon-v1.2.0-build20240327.tar.gz
    a edge-release-images-tgz-20240705/hardened-sriov-cni-v1.2.0-build20240327.tar.gz
    a edge-release-images-tgz-20240705/hardened-ib-sriov-cni-v1.2.0-build20240327.tar.gz
    a edge-release-images-tgz-20240705/hardened-sriov-network-device-plugin-v1.2.0-build20240327.tar.gz
    a edge-release-images-tgz-20240705/hardened-sriov-network-resources-injector-v1.2.0-build20240327.tar.gz
    a edge-release-images-tgz-20240705/hardened-sriov-network-webhook-v1.2.0-build20240327.tar.gz
    ```
    
  - Upload your tarball to your private registry to preload with the images downloaded in the previous step using the [following script](https://github.com/suse-edge/fleet-examples/blob/release-3.0/scripts/day2/edge-load-images.sh)

### Building the Edge Cluster Image using EIB

All the following commands in this section could be executed using any linux laptop/server x86_64 with podman installed. You don't need to have a specific environment to build the image.

#### Generate the image with our configuration for Telco profile

```
$ cd telco-examples/edge-clusters

> $ sudo podman run --rm --privileged -it -v $PWD:/eib \

Author: depending on the cgroup version, right?

In case you want to use a dhcp-less environment, you will need to configure and replace also the following parameters:

- `${CONTROLPLANE_INTERFACE}` - The control plane interface to be used for the edge cluster (e.g `eth0`).
- `${CONTROLPLANE_IP}` - The IP address to be used as a endpoint for the edge cluster (should match with the kubeapi-server endpoint).
- `${CONTROLPLANE_PREFIX}` - The CIDR to be used for the edge cluster (e.g `24` in case you want `/24` or `255.255.255.0`).
- `${CONTROLPLANE_GATEWAY}` - The gateway to be used for the edge cluster (e.g `192.168.100.1`).
- `${CONTROLPLANE_MAC}` - The MAC address to be used for the control plane interface (e.g `00:0c:29:3e:3e:3e`).
- `${DNS_SERVER}` - The DNS to be used for the edge cluster (e.g `192.168.100.2`).

Then, you need to apply the changes using the following command into the management cluster:

```
$ kubectl apply -f bmh-example.yaml
```

The new Baremetal host will be enrolled changing the state from registering to inspecting and available. You could check the status using the following command:

``` 
$ kubectl get bmh
```

> #### Provision the new host using the CAPI manifests and the image generated in an air-gap environment

Author: let's change to: Provision the new host using the CAPI manifests

Once the new Baremetal host is available, you need to provision the new host using the CAPI manifests and the image generated in the previous step.

The first thing is to modify the `telco-capi-airgap.yaml` file and replace the following with your values:

- `${TLS_BASE64_CERT}` and `${TLS_BASE64_KEY}` - In the case your private registry contains a client certificate and key, you can include it and need to be encoded them in base64 and replace the values in the `telco-capi-airgap.yaml` file.
- `${CA_BASE64_CERT}` - In the case your private registry enables TLS, at least you need to include the CA certificate to be used for the private registry. You need to encode it in base64 and replace the value in the `telco-capi-airgap.yaml` file.

> - `${REGISTRY_USERNAME}` and `${REGISTRY_PASSWORD}` - In the case your private registry enables the basic-auth mecanism, you need to include the username and password to be used for the private registry. You need to encode it in base64 and replace the values in the `telco-capi-airgap.yaml` file.

Author: it's ok, Base64 encoded
